File: lib/services/firestore_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:firebase_core/firebase_core.dart';
import 'package:flutter/foundation.dart';
import '../models/saved_project.dart';
import '../models/snippet.dart';

class FirestoreService {
  // Safe access to Firebase instances
  bool get isReady {
    try {
      return Firebase.apps.isNotEmpty;
    } catch (_) {
      return false;
    }
  }

  FirebaseFirestore? get _db => isReady ? FirebaseFirestore.instance : null;
  FirebaseAuth? get _auth => isReady ? FirebaseAuth.instance : null;

  Stream<List<Map<String, dynamic>>> getPublicTemplates() {
    final db = _db;
    if (db == null) return Stream.value([]);
    
    return db.collection('public_templates').snapshots().map((snapshot) {
      return snapshot.docs.map((doc) {
        var data = doc.data();
        data['id'] = doc.id;
        return data;
      }).toList();
    });
  }

  // Missing method required by Admin Dashboard
  Future<void> savePublicTemplate({
    required String name,
    required String description,
    required List<Map<String, dynamic>> elements,
  }) async {
    final db = _db;
    if (db == null) return;
    await db.collection('public_templates').add({
      'name': name,
      'description': description,
      'elements': elements,
      'createdAt': FieldValue.serverTimestamp(),
    });
  }

  // Personal Collections
  DocumentReference? get _userDoc {
    final auth = _auth;
    final db = _db;
    if (auth?.currentUser == null || db == null) return null;
    return db.collection('users').doc(auth!.currentUser!.uid);
  }

  Future<void> saveProject(SavedProject project) async {
    final userDoc = _userDoc;
    if (userDoc == null) return;
    await userDoc.collection('projects').doc(project.id).set(project.toJson());
  }

  Stream<List<SavedProject>> getProjects() {
    final userDoc = _userDoc;
    if (userDoc == null) return Stream.value([]);
    
    return userDoc.collection('projects')
        .orderBy('lastModified', descending: true)
        .snapshots()
        .map((snapshot) {
      return snapshot.docs.map((doc) => SavedProject.fromJson(doc.data())).toList();
    });
  }

  Future<void> deleteProject(String id) async {
    final userDoc = _userDoc;
    if (userDoc == null) return;
    await userDoc.collection('projects').doc(id).delete();
  }

  Future<void> saveSnippet(Snippet snippet) async {
    final userDoc = _userDoc;
    if (userDoc == null) return;
    await userDoc.collection('snippets').doc(snippet.id).set(snippet.toJson());
  }

  Stream<List<Snippet>> getSnippets() {
    final userDoc = _userDoc;
    if (userDoc == null) return Stream.value([]);
    
    return userDoc.collection('snippets').snapshots().map((snapshot) {
      return snapshot.docs.map((doc) => Snippet.fromJson(doc.data())).toList();
    });
  }

  Future<void> deleteSnippet(String id) async {
    final userDoc = _userDoc;
    if (userDoc == null) return;
    await userDoc.collection('snippets').doc(id).delete();
  }
}

File: src/services/FirestoreService.js
import { getAuth, getFirestore, firestoreFieldValue } from '../platform/firebase';

class FirestoreService {
  // --- App Configuration (Admin) ---
  getAppConfig(callback) {
    return getFirestore()
      .collection('settings')
      .doc('app_config')
      .onSnapshot(documentSnapshot => {
        callback(documentSnapshot.data() || {});
      });
  }

  async updateAppConfig(data) {
    await getFirestore()
      .collection('settings')
      .doc('app_config')
      .set(data, { merge: true });
  }

  // --- User Management ---
  async updateUserStatus(uid, updates) {
    await getFirestore().collection('users').doc(uid).update(updates);
  }

  // --- Templates ---
  getPublicTemplates(callback) {
    return getFirestore()
      .collection('public_templates')
      .onSnapshot(querySnapshot => {
        const templates = querySnapshot.docs.map(doc => ({
          id: doc.id,
          ...doc.data(),
        }));
        callback(templates);
      });
  }

  async savePublicTemplate(name, description, elements, category = 'General') {
    await getFirestore().collection('public_templates').add({
      name,
      description,
      elements,
      category,
      createdAt: firestoreFieldValue.serverTimestamp(),
    });
  }

  async deleteTemplate(id) {
    await getFirestore().collection('public_templates').doc(id).delete();
  }

  // --- Personal Collections ---
  async saveProject(project) {
    const user = getAuth().currentUser;
    if (!user) return;
    await getFirestore()
      .collection('users')
      .doc(user.uid)
      .collection('projects')
      .doc(project.id)
      .set(project);
  }

  getProjects(callback) {
    const user = getAuth().currentUser;
    if (!user) return () => {};
    return getFirestore()
      .collection('users')
      .doc(user.uid)
      .collection('projects')
      .orderBy('lastModified', 'desc')
      .onSnapshot(querySnapshot => {
        const projects = querySnapshot.docs.map(doc => doc.data());
        callback(projects);
      });
  }
}

export default new FirestoreService();
